docs/dev/core/internal/upload.md

```diff
@@ -7,6 +7,62 @@ The Glean SDK leverages the available platform capabilities to implement any net
 Glean core controls all upload and coordinates the platform side with its own internals.
 All language bindings implement ping uploading around a common API and protocol.
 
+## The upload module in the language bindings
+This paragraph illustrates the recommended structure of the upload module implementation in the language bindings (regardless of the target language), in order to allow greater flexibility with respect to potential future requirements and features.
+
+```mermaid
+classDiagram
+    class UploadResult {
+        ~ToFFI()* int
+    }
+
+    class HttpResponse {
+        int statusCode
+        ~ToFFI() int
+    }
+
+    class UnrecoverableFailure {
+        ~ToFFI() int
+    }
+
+    class RecoverableFailure {
+        ~ToFFI() int
+    }
+
+    class PingUploader {
+        <<interface>>
+        +Upload() UploadResult
+    }
+
+    class BaseUploader {
+        +BaseUploader(PingUploader)
+        ~TriggerUploads()
+        ~CancelUploads()
+    }
+
+    class HttpUploader {
+        +Upload() UploadResult
+    }
+
+    UploadResult <|-- HttpResponse
+    UploadResult <|-- UnrecoverableFailure
+    UploadResult <|-- RecoverableFailure
+    PingUploader <|-- HttpUploader
+    PingUploader o-- BaseUploader
+
+```
```

Comment:
Is there a way to provide a text caption to this figure?

Reply:
Not that I know of, I'm afraid :(

Reply:
Since markdown accepts any valid HTML, you could put this inside a [`figure`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/figure) tag and provide a `figcaption` right next to it. Apparently mermaid creates an `img` tag where all this code is, so that would work perfectly. I could be wrong as to how mermaid works though.

Reply:
This doesn't seem to work, at least locally, for me :\ The mermaid doesn't seem to get parsed?

Reply:
Hum, oh well. Maybe we can open a follow up to look into this with more time. I see the other upload graph we have also doesn't have a caption so it could be a more broad thing where we add captios to all images in the book.

Reply:
Filed https://bugzilla.mozilla.org/show_bug.cgi?id=1652705